adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -391,9 +404,17 @@ public AuthenticationResult getToken(String authorizationUrl)
 
                 // Check if we have code
                 if (result != null && result.getCode() != null && !result.getCode().isEmpty()) {
+                    if (!StringExtensions.isNullOrBlank(result.getAuthority())) {
+                        setTokenEndpoint(result.getAuthority() + DEFAULT_TOKEN_ENDPOINT);
```

Comment:
I wouldn't do the setTokenEndpoint in different places, mTokenEndpoint is not a final variable, it's not guaranteed to be set in the class, there would be possibilities that it's null due to bug. When you construct the token endpoint, the only variable is the cloud instance, your set method can pass in the instance. Construct the token endpoint in the set method and set it in one place.